### ISSUE TITLE
Bridge.Console improvements

### ISSUE THREAD

**Issue by geoffreymcgill** [bridgedotnet/Bridge#1994]
The **Bridge.Console** utility needs some improvements. Ideas are being collected here.

### Improvements
* [x] Make optional by removing **bridge.console.js** `<script>` from .html
**Bridge Console** is emitted as a separate **bridge.console.js** which can be added to the applications .html. To prevent **Bridge Console** from rendering, just remove **bridge.console.js**. See https://github.com/bridgedotnet/Bridge/issues/1994#issuecomment-296527030 and https://github.com/bridgedotnet/Bridge/issues/1994#issuecomment-296754941
* [x] Capture JavaScript errors
* [x] Render JavaScript Exception messages (#2613)
**Bridge Console** catches all JavaScript errors and exceptions, then display with some client-side stack details ([Deck](https://deck.net/748dc599b769246f20fe58dc74919464))
* [x] Better rendering of custom `[object]` objects in browser console (#2556)
* [x] Rendering `JSON.stringify` value (instead of its type name) for objects having no `ToString()`  method overridden ([Deck](https://dev.deck.net/08d89168103b2058fb0eebf4d396e116)). Note this is not a **.Net** standard behavior
* [x] Multi-line mode (next output in after `Console.WriteLine()` will start from a new line in **Bridge.Console**)

### Fixes

* [x] Support `Console.WriteLine(char[])` and `Console.WriteLine(char[], int, int)`  (#2617)
* [x] Support `Console.Write(char[])` and `Console.Write(char[], int, int)`  (#2617)
* [x] Support `System.Console.Clear()` (#2618)
* [x] Incorrect formatting in `Console.Write(double)` (#2619)

## See Also

* #1993
* #2556
* #2613
* #2617
* #2618
* #2619

**Comment by DanTup** [comment]
I added [a comment](https://github.com/bridgedotnet/Bridge/issues/1993#issuecomment-256591824) in #1993 that maybe more relevant here:

> It seems a little strange to me to ship all the console code inside bridge.json and have another package to disable it. A switch to toggle is better, but it does feel a bit bloated to include a dev feature in all production builds that you might not want. Wouldn't it be better to have it in a separate package that you pull in if you want it, but otherwise it's not there?

**Comment by DanTup** [comment]
@geoffreymcgill Do you still plan to let us opt-out of this console? It's rather annoying to have this weird in-page console pop up when I have the browser console open (and can do more in there).

**Comment by geoffreymcgill** [comment]
Yes, we will try to add some settings to disable **Bridge Console** in Bridge 16.0.

How about using the following to write to the browser console directly if you want to completely avoid **Bridge Console**?

http://deck.net/bb9e377ca599b32827d9cbf0954f3c34

```csharp
//@ console.log(msg);
```

**Comment by DanTup** [comment]
Thanks! I've been using @ProductiveRage's workaround for now (so I can still use the C# Console methods), it'd just be nice to have a cleaner (and supported) way :-)

**Comment by geoffreymcgill** [comment]
A pull request has been submitted with **Bridge.Console** improvements. Once testing is finished, the PR will be merged into the **master** branch and will be included with the next Bridge release. 

The new functionality can be tested at https://dev.deck.net.

**Bridge.Console** has been separated into a dedicated **bridge.console.js** file and is now completely optional by adding or removing the .js file from the .html. 

```html
<!DOCTYPE html>

<html lang="en" xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta charset="utf-8" />
    <title>Bridge ClassLibrary28</title>
    <script src="bin/Debug/bridge/bridge.min.js"></script>

    <!-- 
            Q: Don't want Bridge.Console? 

            A: Remove bridge.console.js from this file
    -->
    <script src="bin/Debug/bridge/bridge.console.min.js"></script>

    <script src="bin/Debug/bridge/demo.js"></script>
</head>
<body>
    <!--
        Right-Click on this file and select "View in Browser"
    -->
</body>
</html>
```

Hope this helps.

**Comment by DanTup** [comment]
Great! Is this still separate/removable even if you use `combineScripts`?

**Comment by geoffreymcgill** [comment]
The **bridge.console.js** is automatically emitted to the output folder and will be combined if `"combineScript": true`.

For better control over combining the .js files, please use [resources](https://github.com/bridgedotnet/Bridge/wiki/global-configuration#resources) config.

With the `resources` config, you can explicitly exclude files from being emitted to the `output` folder. Please see the following sample for excluding **bridge.console.js**:

```js
{
  "resources": [
    {
      "name": "bridge.console.js",
      "extract": false
    }
  ]
}
```

Hope this helps.

**Comment by DanTup** [comment]
Great, thanks!
